### PR TITLE
feat: interactive PTY support across E2B, NodeJS bridge, Tauri

### PR DESCRIPTION
**Pull request by fkesheh**
## Summary

Extends `run_terminal_cmd` with an interactive PTY path so the model can drive REPLs, prompts, sudo, SSH, installers, etc. — anything `sandbox.commands.run()` can't.

- New `action` dispatcher: `exec | send | wait | view | kill`. Legacy callers with `{command, explanation, is_background, timeout}` still work byte-for-byte (action defaults to `exec`, interactive defaults to `false`).
- Per-chat `PtySessionManager` with ring buffer, idle / lifetime timers, concurrency cap (2 per chat), auto-cleanup on `onFinish`.
- Backends:
  - **E2B** via `sandbox.pty.*` (`e2b-pty-adapter.ts`).
  - **NodeJS bridge** & **Tauri desktop** via Centrifugo pub/sub (`centrifugo-pty-adapter.ts`) + local runner commands.
  - Rust `portable-pty` native path in `packages/desktop/src-tauri/src/pty.rs` for Tauri.
- UI already rendered via pre-existing `ShellAction` chips + `data-terminal` stream parts (`TerminalToolHandler`, `ComputerSidebar`, `TerminalCodeBlock`). Rendering uses `@xterm/headless` so cursor/erase CSI sequences produce correct text instead of garbling.
- `waitForOutput` pre-captures buffered bytes before subscribing so E2B's RTT doesn't miss the first echo burst.
- Tauri fix: publish queue serializes `pty_data` emissions so rapid per-read chunks preserve FIFO order (prevents garbled terminal rendering under load).

## PR review fixes (latest commit `9fc2f68f`)

- Centrifugo `kill()` now resolves `exited` before tearing down the subscription so callers don't stall on a `pty_exit` that can't arrive post-cleanup. Cleanup + exited-resolution are idempotent.
- Desktop bridge: `pty_resize` / `pty_kill` invoke failures log instead of silently swallowing.
- Rust reader thread breaks out cleanly when the Tauri `Channel` closes (window gone / subscription dropped) instead of looping and dropping bytes.
- Dropped orphan `PtyHandle` re-export.
- Added FIFO-ordering regression test for the Tauri publish-queue fix.

## Test plan

- [x] 880 unit tests pass (`pnpm test`).
- [x] `pnpm typecheck` clean.
- [x] Manual E2B: `python3` REPL, `npm init` interactive flow, `send` keystrokes, `view` snapshot, `kill`.
- [x] Manual Tauri: same flows on macOS via portable-pty; verified FIFO output after publish-queue fix.
- [x] Manual NodeJS bridge via Centrifugo relay.
- [x] Regression check: existing non-interactive `run_terminal_cmd` callers unchanged.
- [ ] Verify Linux Tauri build.
- [ ] Verify Windows Tauri build (conpty behavior).

## Out of scope / follow-ups

- Cross-turn session resumption (Convex-persisted `chat_pty_sessions`, E2B `.connect(pid)` on reload).
- User-consent UI for first PTY spawn on desktop.
- `reason` / `error` fields on result shape (model-facing contract — wants explicit sign-off).
- Centrifugo adapter unit tests (currently exercised only via integration).

## Summary by CodeRabbit

* **New Features**
  * Interactive terminal sessions: exec, send, wait (pattern/idle), view snapshots, and kill; streaming now includes action + session metadata and session snapshots.
* **UI**
  * Pulsing "executing" indicator; sidebar shows PID or short session ID and clearer interactive vs. standard terminal labels; terminal formatting and loading text adapt for interactive actions.
* **Reliability**
  * Robust per-chat PTY session manager, ordered PTY data delivery, and best-effort teardown on stream finish/error/abort.
* **Tests**
  * Extensive test coverage for adapters, session manager, interactive flows, and edge cases.
* **Chores**
  * Desktop/local PTY backends and headless terminal rendering added.

### CODE REVIEW

File: app/services/desktop-sandbox-bridge.ts [hunk 1]
```diff
@@ -275,6 +312,145 @@ export class DesktopSandboxBridge {
...
+        pid: result.pid,
+      });
+    } catch (err) {
+      await this.publishResult({
+        type: "pty_error",
+        sessionId,
+        message: err instanceof Error ? err.message : String(err),
+      });
+    }
+  }
```
[the remark this answers is not in the document]
Comment (fkesheh) [reply app/services/desktop-sandbox-bridge.ts:475]: Fixed in 5525d75c: `pty_ready` now goes through `enqueuePublish` so it can't arrive ahead of already-queued `pty_data` chunks. The catch branch keeps `publishResult` direct because the error path never reaches the `channel.onmessage` listener (no `pty_data` queued) and `enqueuePublish` is out of scope there.

File: lib/ai/tools/utils/pty-session-manager.ts [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+          (info) => {
+            session.exitedNaturally = { exitCode: info.exitCode };
+          },
+          () => {
+            session.exitedNaturally = { exitCode: null };
+          },
+        )
+        .catch((err) =>
+          console.error("[pty-session-manager] exited handler failed:", err),
+        );
```
[the remark this answers is not in the document]
Comment (fkesheh) [reply lib/ai/tools/utils/pty-session-manager.ts:163]: Noted — tracking as a follow-up rather than blocking this PR. The cap was bumped from 2 to 10 earlier in this branch, which defangs the short-term impact. A proper post-exit reaper (e.g. `POST_EXIT_IDLE_MS` or excluding `exitedNaturally` sessions from the concurrency count while keeping the buffer available for `view`/`wait`) deserves its own PR with targeted tests.

File: lib/ai/tools/run-terminal-cmd.ts [hunk 3]
```diff
@@ -105,37 +324,390 @@ If you are generating files:
...
+          ? translateInputSequence(input)
+          : translateInput(input);
+        if (bytes.byteLength > MAX_INPUT_BYTES_PER_SEND) {
+          return errorResult(
+            `Input exceeds MAX_INPUT_BYTES_PER_SEND=${MAX_INPUT_BYTES_PER_SEND} (got ${bytes.byteLength}).`,
+          );
+        }
+        try {
+          await session.handle.sendInput(bytes);
+        } catch (err) {
```
[the remark this answers is not in the document]
Comment (fkesheh) [reply lib/ai/tools/run-terminal-cmd.ts]: Intentionally declined. `action=send` is raw keystrokes for interactive sessions — REPL input, password prompts, single control chars, arrow navigation. Command-regex guardrails are designed for whole shell commands and would false-positive on ordinary REPL text (e.g. typing `rm` into a python REPL as a variable name, or the user's password happening to match a pattern). The tool `describe` already warns explicitly: *"Raw keystrokes BYPASS command guardrails; never paste untrusted content."* The correct threat model is: guardrail the shell command in `action=exec`, accept that once a shell is open via PTY the user is in a shell.
